pkg/api/write.go

```diff
@@ -158,32 +165,60 @@ func decodeSnappy(w http.ResponseWriter, r *http.Request) bool {
 	if detectSnappyStreamFormat(buf) {
 		r.Body = &readCloser{
 			reader: snappy.NewReader(mr),
-			closer: r.Body,
+			closer: originalBody,
 		}
 		return true
 	}
 
-	compressed, err := ioutil.ReadAll(mr)
+	compressed := compressedBufPool.Get().(*bytes.Buffer)
+	compressed.Reset()
+	defer compressedBufPool.Put(compressed)
+
+	_, err = compressed.ReadFrom(mr)
 	if err != nil {
 		invalidRequestError(w, "request body read error", err.Error(), metrics)
 		return false
 	}
 
+	b := decodedBufPool.Get().(*bytes.Buffer)
+	b.Reset()
+	n, err := snappy.DecodedLen(compressed.Bytes())
+	if err != nil {
+		invalidRequestError(w, "snappy decode length error", err.Error(), metrics)
+		return false
+	}
+	b.Grow(n)
+
 	// Snappy block format.
-	decoded, err := snappy.Decode(nil, compressed)
+	decoded, err := snappy.Decode(b.Bytes()[:n], compressed.Bytes())
 	if err != nil {
 		invalidRequestError(w, "snappy decode error", err.Error(), metrics)
 		return false
 	}
 
-	body := bytes.NewBuffer(decoded)
+	newBuf := bytes.NewBuffer(decoded)
 	r.Body = &readCloser{
-		reader: body,
-		closer: r.Body,
+		reader: newBuf,
+		closer: funcCloser(func() error {
+			decodedBufPool.Put(newBuf)
+			return originalBody.Close()
+		}),
 	}
 	return true
 }
 
+var compressedBufPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+var decodedBufPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
```

Comment:
https://morioh.com/p/32c938382e23 makes me think having different pools for objects of different sizes is better. Anyway that could be a separate PR with different benchmarking/analysis.

Reply:
I gave a quick look, how is that linked to our problem?

Our case is a type which is being used at 3 (2 for now, but can be helped in 3 places in total) places. In our case, a bigger slice cap would be helpful in all places, since in net total, lower allocs will be made. If different, the net alloc will be higher, but the biggest would be concentrated to a single pool, as it's happening now. But since there is no guarantee that the ssmaller concentrated ones will never be big in coming samples batch, I think it will be more safer to have bigger cap allocs in net. But yes, a benchmark will make us most confident here.